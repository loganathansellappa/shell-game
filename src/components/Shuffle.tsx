import React, { Children, useEffect, useLayoutEffect, useState } from 'react';
import usePrevious from '../CustomHooks/usePrevious';
import calculateShellBoxes from '../utils/CalculateShellBoxes';

const Shuffle = ({ children }) => {
  const [ShellBox, setShellBox] = useState({});
  const [prevShellBox, setPrevShellBox] = useState({});
  const prevChildren = usePrevious(children);
  useLayoutEffect(() => {
    const newShellBox = calculateShellBoxes(children);
    setShellBox(newShellBox);
  }, [children]);

  useLayoutEffect(() => {
    const prevShellBox = calculateShellBoxes(prevChildren);
    setPrevShellBox(prevShellBox);
  }, [prevChildren]);

  useEffect(() => {
    const hasPrevShellBox = Object.keys(prevShellBox).length;

    if (hasPrevShellBox) {
      Children.forEach(children, (child) => {
        const domNode = child.ref.current;
        const firstBox = prevShellBox[child.key];
        const lastBox = ShellBox[child.key];
        const changeInX = firstBox.left - lastBox.left;

        if (changeInX) {
          requestAnimationFrame(() => {
            domNode.style.transform = `translateX(${changeInX}px)`;
            domNode.style.transition = 'transform 0s';

            requestAnimationFrame(() => {
              domNode.style.transform = '';
              domNode.style.transition = 'transform 500ms';
            });
          });
        }
      });
    }
  }, [ShellBox, prevShellBox, children]);

  return children;
};

export default Shuffle;
