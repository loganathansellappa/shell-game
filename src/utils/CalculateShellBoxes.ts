import { Children } from 'react';

const calculateShellBoxes = (children) => {
  const boundingBoxes = {};

  Children.forEach(children, (child) => {
    const domNode = child.ref.current;
    if (domNode) {
      const nodeBoundingBox = domNode.getBoundingClientRect();

      boundingBoxes[child.key] = nodeBoundingBox;
    }
  });

  return boundingBoxes;
};

export default calculateShellBoxes;
