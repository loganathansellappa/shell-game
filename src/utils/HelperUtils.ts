import { ColorCodes } from '~/@types/GenericTypes';

export const shellCount = (): number => Number(process.env.SHELL_COUNT);

export const shuffle = <T>(elements: T[]): Array<T> => {
  for (let i = elements.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [elements[i], elements[j]] = [elements[j], elements[i]];
  }
  return [...elements];
};

export const getColors = (count: number): ColorCodes => {
  let colorCodes = {};
  for (let i = 0; i < count; i++) {
    colorCodes[`key${i}`] = `#${Math.floor(Math.random() * 16777215).toString(
      16
    )}`;
  }
  return colorCodes;
};

export enum Level {
  Beginner = 2,
  Intermediate = 4,
  Expert = 8,
}
