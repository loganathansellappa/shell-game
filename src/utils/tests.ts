import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { BrowserRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';

export const renderWithRouter = (ui: JSX.Element, { route = '/' } = {}) => {
  window.history.pushState({}, 'Test page', route);

  return {
    user: userEvent.setup(),
    ...render(ui, { wrapper: BrowserRouter }),
  };
};
export const testStore = () => {
  const mockStore = configureStore();
  const initialState = {
    gameLog: {
      gameCount: 1,
      winners: 1,
      losers: 0,
    },
  };
  const store = mockStore(initialState);
  return store;
};
