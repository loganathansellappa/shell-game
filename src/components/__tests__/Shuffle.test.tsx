import { renderWithRouter } from '../..//utils/tests';
import Shuffle from '../Shuffle';
import { createRef } from 'react';

test('renders Shuffle component', () => {
  const { container } = renderWithRouter(
    <Shuffle>
      <div ref={createRef()}>test</div>
    </Shuffle>
  );
  expect(container.firstChild).toMatchSnapshot();
});
