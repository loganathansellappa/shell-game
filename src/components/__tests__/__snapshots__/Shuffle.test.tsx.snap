// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Shuffle component 1`] = `
<div>
  test
</div>
`;
